### PR TITLE
Group connected variables together Closes #40

### LINKED ISSUE #40: Group naturally connected/depended variables together
### Description

Join variables into _structs_ in places where it is applicable and make sense from a clean code perspective. With this technique a number of functions parameters could be smaller.

- if some variables are naturally connected/depended between each other it would be better to join them into a _struct_
- an example of joining two connected variables:
```
// instead of having two independent variables
let key = vec![1,2,3,4];
let value = vec![5,6,7,8];

// it is better to join them together into a struct
struct KVPair(Vec<u8>, Vec<u8>);
let kv_pair = KVPair(vec![1,2,3,4], vec![5,6,7,8]);
```
- check if any _struct_ could be present as a field in some other _struct_
- check if any _struct_ could be joined with some other _struct_ into a new struct

### Acceptance Criteria

- variables are joined into _structs_ or _tuples_ where it makes sense
- different _structs_ don't have the same fields present inside them but in this case one _struct_ is embedded inside another _struct_ 

### Additional Information

- possible candidates for this type of refactorizations are `QueryProof`, `QueryProofWithProof` and `Proof` structs
- In a lot of places in the code we have key-value pairs. Maybe join them in a _tuple_ or _struct_?

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #40 

### How was it solved?

- Make some new structs to group connected variables
- Make some new strong types for increase readability
- Remove duplicate codes

### How was it tested?

All tests passed.

### CODE REVIEW

File: src/smt.rs [hunk 1]
```diff
@@ -447,13 +442,39 @@ impl rocksdb::WriteBatchIterator for UpdateData {
     }
 }
 
+impl QueryProof {
```
Comment (matjazv) [inline src/smt.rs:445]: The same applies here regarding inline functions.

**Review by matjazv — CHANGES_REQUESTED** [review]
Thank you for adding some comments in the code.

Consider squashing commits to have only 1-3 of them.

File: src/consts.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
+use crate::types::{KeyLength, SubtreeHeight, SubtreeHeightKind};
+
 pub const PREFIX_STATE: &[u8] = &[0];
 pub const PREFIX_SMT: &[u8] = &[1];
 pub const PREFIX_DIFF: &[u8] = &[2];
 
-pub const KEY_LENGTH: usize = 38;
-pub const SUBTREE_SIZE: usize = 8;
+pub const KEY_LENGTH: KeyLength = KeyLength(38);
+pub const SUBTREE_SIZE: SubtreeHeight = SubtreeHeight(SubtreeHeightKind::Eight);
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/consts.rs:8]: I think `SUBTREE_HEIGHT` is better.

File: src/types.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+    pub fn sub_to_usize(self, value: u8) -> usize {
+        (self.0 - value as u16) as usize
+    }
+
+    pub fn div_to_usize(self, value: u16) -> usize {
+        (self.0 / value) as usize
+    }
+
+    pub fn rem_to_u8(self, value: u16) -> u8 {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:207]: I used that name based on operator `%` and its trait name. `rem` as remaining value.

File: src/types.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    fn new() -> Self;
+}
+
+pub trait KVPairCodec {
+    fn decode(val: &[u8]) -> Result<KVPair, codec::CodecError>;
+    fn encode(&self) -> Vec<u8>;
+}
+
+impl New for Cache {
+    fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:58]: add `#[inline(always)]` for each small functions of `types.rs` could be a good increase performance.

File: src/types.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    pub fn rem_to_u8(self, value: u16) -> u8 {
+        (self.0 % value) as u8
+    }
+
+    pub fn to_be_bytes(self) -> [u8; 2] {
+        self.0.to_be_bytes()
+    }
+
+    // Cast to u32 and returns with len(4) for JS API
+    pub fn as_u32_to_be_bytes(self) -> [u8; 4] {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:216]: `be` represented as `big-endian`

File: src/types.rs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+    pub fn div_to_usize(self, value: u16) -> usize {
+        (self.0 / value) as usize
+    }
+
+    pub fn rem_to_u8(self, value: u16) -> u8 {
+        (self.0 % value) as u8
+    }
+
+    pub fn to_be_bytes(self) -> [u8; 2] {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:211]: `be` represented as `big-endian`

File: src/types.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:207]: Done.

File: src/types.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+pub type Cache = HashMap<Vec<u8>, Vec<u8>>;
+pub type VecOption = Option<Vec<u8>>;
+
+// Strong type of SMT with max value KEY_LENGTH * 8
+#[derive(Clone, Debug, Copy, PartialEq, Eq)]
+pub struct Height(pub u16);
+
+// Strong type of structure position in Subtree with max value 2 ^ SUBTREE_SIZE
+#[derive(Clone, Debug, Copy, PartialEq, Eq)]
+pub struct StructurePostion(pub u16);
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:16]: Done.

File: src/smt.rs [hunk 8]
```diff
@@ -682,13 +698,13 @@ impl SubTree {
         Self {
             structure,
             nodes: node_hashes,
-            root: empty.hash,
+            root: empty.hash.value_as_vec(),
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/smt.rs:701]: The previous hash was a `Vec` but the new one is a key-value pair of `KVPair` and we want to move the value of the pair outside of the box.

File: src/smt.rs [hunk 9]
```diff
@@ -464,7 +485,7 @@ impl UpdateData {
         Self { data: new_data }
     }
 
-    pub fn entries(&self) -> (Vec<Vec<u8>>, Vec<Vec<u8>>) {
+    pub fn entries(&self) -> (NestedVec, NestedVec) {
         let mut kvpairs = vec![];
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/smt.rs:489]: Done.

File: src/types.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:58]: The result of the different runs:

1. Normal code without tag : `42.42s`
2. With `inline(always)` tag: `42.69s`
3. with `inline` tag: `41.92s`

I tried to run 4 times and the smallest time was `40.36s` for `inline` tag, so add `inline` to improve performance.

File: src/smt.rs (hunk 1 above)
Comment (hrmhatef) [reply src/smt.rs:445]: use `inline` tag that explained in above comment.

**Review by matjazv — APPROVED** [review]

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: src/types.rs [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+    #[inline]
+    pub fn is_empty_value(&self) -> bool {
+        self.1.is_empty()
+    }
+}
+
+impl Height {
+    #[inline]
+    pub fn is_equal_with(self, value: u16) -> bool {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/types.rs:211]: Done.

**Review by shuse2 — APPROVED** [review]

**Pull request merged**